src/main/java/org/opensrp/web/config/security/filter/XssPreventionRequestWrapper.java

```diff
@@ -0,0 +1,129 @@
+package org.opensrp.web.config.security.filter;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.JsonNodeType;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.apache.commons.io.IOUtils;
+import org.owasp.encoder.Encode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.ServletInputStream;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.Map;
+
+public class XssPreventionRequestWrapper extends HttpServletRequestWrapper {
+
+	private static Logger logger = LoggerFactory.getLogger(XssPreventionRequestWrapper.class);
+
+	private static ObjectMapper mapper = new ObjectMapper();
+
+	private byte[] rawData;
+
+	private HttpServletRequest request;
+
+	private ResettableServletInputStream servletStream;
+
+	public XssPreventionRequestWrapper(HttpServletRequest request) {
+		super(request);
+		this.request = request;
+		this.servletStream = new ResettableServletInputStream();
+	}
+
+	public void resetInputStream() {
+		servletStream.stream = new ByteArrayInputStream(rawData);
+	}
+
+	@Override
+	public ServletInputStream getInputStream() throws IOException {
+		if (rawData == null) {
+			rawData = IOUtils.toByteArray(this.request.getReader());
+			servletStream.stream = new ByteArrayInputStream(rawData);
+		}
+		updateParameters();
+		return servletStream;
+	}
+
+	@Override
+	public BufferedReader getReader() throws IOException {
+		if (rawData == null) {
+			rawData = IOUtils.toByteArray(this.request.getReader());
+			servletStream.stream = new ByteArrayInputStream(rawData);
+		}
+		updateParameters();
+		return new BufferedReader(new InputStreamReader(servletStream));
+	}
+
+	private void updateParameters() throws IOException {
+		String requestBody = new String(rawData, StandardCharsets.UTF_8);
+		if (isValidJSON(requestBody)) {
+			JsonNode jsonNode = mapper.readTree(requestBody);
+			JsonNode updatedJsonNode = encode(jsonNode);
+			String encodedJsonString = updatedJsonNode.toString();
+			rawData = encodedJsonString.getBytes();
+			servletStream.stream = new ByteArrayInputStream(rawData);
+		}
+	}
+
+	private static JsonNode encode(JsonNode node) {
+		if (node.isValueNode()) {
+			if (JsonNodeType.STRING == node.getNodeType() && !isEscapedString(node)) {
+				return JsonNodeFactory.instance.textNode(Encode.forJava(node.asText()));
```

Comment:
Use `org.owasp.encoder.Encode.forHtmlContent(String)` for encoding of any HTML characters. Using `Encode.forJava` are still saved. I was able to save `{"id":2,"name":"Facility","active":true,"description":"Health Facility <script src=\"http://google.com\"/>"}` with the HTML tags in place